builder/builder-main.c

```diff
@@ -408,7 +411,7 @@ main (int    argc,
 
   builder_manifest_checksum (manifest, cache, build_context);
 
-  if (!builder_cache_lookup (cache, "init"))
+  if (!opt_finish_only && !builder_cache_lookup (cache, "init"))
```

Comment:
Can you put this and the below opt_finish_only check into s single if that wraps both the init and build phases, similar to the opt_build_only case is handled a bit below.